QccPack-0.61-1/src/libQccPackECC/utils/man1/rsencode.1
.TH RSENCODE 1 "QCCPACK" ""
.SH NAME
rsencode \- Reed-Solomon coding
.SH SYNOPSIS
.B rsencode
.RB "[\|" \-k
.IR  k "\|]"
.RB "[\|" \-n
.IR  n "\|]"
.I infile
.I outfile
.SH OPTIONS
.TP
.BI \-k " k"
Integer. The number of message symbols per coded block.
Default = 10.
.TP
.BI \-n " n"
Integer. The total number of symbols per coded block.
Default = 15.
.SH DESCRIPTION
.LP
.B rsencode
implements Reed-Solomon forward-error-correction (FEC) coding
of a file.
.I infile
is the filename of the file to be FEC protected;
.I outfile
is the resulting FEC-protected file.
It is assumed that symbols stored in
.I infile
are single-character bytes; i.e., it is a text file.
.LP
Reed-Solomon coding is a
.RI ( n ", " k )
FEC code; that is,
.IR n " - " k
FEC symbols are concatenated to
.I k
message symbols
to produce an
.IR n -symbol
block. The 
.IR n " - " k
FEC symbols are redundant information that aids in the recovery
of the
.I k
message symbols should some of these symbols be lost or corrupted
during transmission of the coded file.
In particular, Reed-Solomon coding has the property that if up to
.IR n " - " k
symbols out of the 
.I n
symbols in the coded block are lost, the original
.I k
message symbols can be recovered exactly.
.LP
.B rsencode
reads
.I k
symbols (i.e., single-byte characters) from
.IR infile ,
calculates the corresponding
.IR n " - " k
FEC symbols for these
.I k
message symbols, and then
writes the corresponding 
.RI length- n
coded block to
.IR outfile .
When the end of file of
.I infile
is encountered, zeros are added as needed to fill out the last block
to 
.I k
symbols before calculating the FEC symbols.
Thus the length of
.I outfile
is
.RI ceil( N / k )* n ,
where 
.I N
is the length of
.IR infile .
.SH NOTES
Reed-Solomon codes were introduced in 1960 by Reed and Solomon. The
implementation found in QccPack follows that by Rizzo.
.SH "SEE ALSO"
.BR QccECCReedSolomonEncode (3),
.BR QccECCReedSolomonDecode (3),
.BR QccPackECC (3),
.BR QccPack (3)

I. S. Reed and G. Solomon,
"Polynomial Codes Over Certain Finite Fields,"
.IR "J. Soc. Ind. Appl. Math." ,
vol. 8, pp. 300-304, 1960. 

L. Rizzo, "Effective Erasure Codes for Reliable Computer Communication
Protocols,"
.IR "ACM Computer Communication Review" ,
vol. 27, no. 2, pp. 24-36, April 1997.
.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.



QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackECC/utils/rsencode.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "rsencode.h"

#define USG_STRING "[-k %d:k] [-n %d:n] %s:infile %s:outfile"


QccString InfileName;
QccString OutfileName;


int main(int argc, char *argv[])
{
  QccECCReedSolomonCode code;
  QccECCFieldElement *message = NULL;
  QccECCFieldElement *coded_message = NULL;
  FILE *infile = NULL;
  FILE *outfile = NULL;
  int index;
  char current_ch;

  QccInit(argc, argv);

  QccECCReedSolomonCodeInitialize(&code);

  /*  Default: (15, 10) code - 10 messages symbols  */
  /*  protected by 5 FEC symbols  */
  code.message_length = 10;
  code.code_length = 15;

  /*  A message alphabet of 256 symbols  */
  code.field.size_exponent = 8;

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &code.message_length,
                         &code.code_length,
                         InfileName,
                         OutfileName))
    QccErrorExit();

  if ((message =
       (QccECCFieldElement *)malloc(sizeof(QccECCFieldElement) *
                                    code.message_length)) == NULL)
    {
      QccErrorAddMessage("%s: Error allocating memory");
      QccErrorExit();
    }

  if ((coded_message =
       (QccECCFieldElement *)malloc(sizeof(QccECCFieldElement) *
                                    code.code_length)) == NULL)
    {
      QccErrorAddMessage("%s: Error allocating memory");
      QccErrorExit();
    }

  if (QccECCReedSolomonCodeAllocate(&code))
    {
      QccErrorAddMessage("%s: Error calling QccECCReedSolomonCodeAllocate()");
      QccErrorExit();
    }

  if ((infile = QccFileOpen(InfileName, "r")) == NULL)
    {
      QccErrorAddMessage("%s: Error calling QccFileOpen()");
      QccErrorExit();
    }

  if ((outfile = QccFileOpen(OutfileName, "w")) == NULL)
    {
      QccErrorAddMessage("%s: Error calling QccFileOpen()");
      QccErrorExit();
    }

  index = 0;
  while (!feof(infile))
    {
      for (index = 0; index < code.message_length; index++)
        if (QccFileReadChar(infile, &current_ch))
          message[index] = 0;
        else
          message[index] = (QccECCFieldElement)current_ch;

      if (QccECCReedSolomonEncode(message, coded_message, &code))
        {
          QccErrorAddMessage("%s: Error calling QccECCReedSolomonEncode()");
          QccErrorExit();
        }

      for (index = 0; index < code.code_length; index++)
        if (QccFileWriteChar(outfile,
                             (char)coded_message[index]))
        {
          QccErrorAddMessage("%s: Error calling QccFileWriteChar()");
          QccErrorExit();
        }
    }

  QccFileClose(infile);
  QccFileClose(outfile);

  QccECCReedSolomonCodeFree(&code);

  QccExit;
}
